### ISSUE TITLE
[Bug] Crash when calling `tiles_changed` in some circumstances

### ISSUE THREAD

**Issue by PatrickLerner** [mich101mich/hierarchical_pathfinding#5]
**Describe the bug**

When calling `tiles_changed`, the library crashes in some certain circumstances with an error like shown below.

**To Reproduce**

Minimum working example. It is random because I don't understand what logic exactly leads to the crash, but it crashes every time I run it and I would expect it to never crash. Note I am not even generating a path.

```rust
use hierarchical_pathfinding::prelude::*;
use rand::{thread_rng, Rng};

const COST_MAP: [isize; 3] = [1, 10, -1];
type Grid = [[i32; 5]; 5];

// only borrows the Grid when called
fn cost_fn(grid: &Grid) -> impl '_ + FnMut((usize, usize)) -> isize {
    move |(x, y)| COST_MAP[grid[y as usize][x as usize] as usize]
}

fn main() {
    let mut grid = [
        [0, 2, 0, 0, 0],
        [0, 2, 2, 2, 0],
        [0, 1, 0, 0, 0],
        [0, 1, 0, 2, 0],
        [0, 0, 0, 2, 0],
    ];
    let (width, height) = (grid.len(), grid[0].len());

    let mut pathfinding = PathCache::new(
        (width, height),
        cost_fn(&grid),
        ManhattanNeighborhood::new(width, height),
        PathCacheConfig {
            chunk_size: 2,
            ..Default::default()
        },
    );

    let mut rand = thread_rng();
    for i in 0..1000 {
        let x = rand.gen_range(0..width);
        let y = rand.gen_range(0..height);
        grid[x][y] = rand.gen_range(0..2);
        pathfinding.tiles_changed(&[(x, y)], cost_fn(&grid));
        println!("iterate {}", i);
    }
}
```

with

```
hierarchical_pathfinding = "0.4"
rand = "0.8"
```

It also fails on master.

**Expected behavior**

Never crash when calling tile updates.

**Actual behavior**

```
thread 'Compute Task Pool (1)' panicked at 'attempt to subtract with overflow', /Users/patrick/.cargo/git/checkouts/hierarchical_pathfinding-4539d4dc23d64d2e/ae726be/src/path/
generic_path.rs:48:19
stack backtrace:
   0:        0x105a51fa4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1920c5bf118add89
   1:        0x105ab66be - core::fmt::write::h9f94d8dd4bbd6f02
   2:        0x105a45e3a - std::io::Write::write_fmt::hcf11ceebd2110ba2
   3:        0x105a55c7f - std::panicking::default_hook::{{closure}}::h25d47083c1b1e6a3
   4:        0x105a55780 - std::panicking::default_hook::heb897de7705ed8da
   5:        0x105a563b5 - std::panicking::rust_panic_with_hook::h7a527f77d3e47d53
   6:        0x105a55f39 - std::panicking::begin_panic_handler::{{closure}}::h03f8ad2b697d744c
   7:        0x105a52418 - std::sys_common::backtrace::__rust_end_short_backtrace::h0847a51f11834207
   8:        0x105a55eca - _rust_begin_unwind
   9:        0x105ae037f - core::panicking::panic_fmt::h3442256188b25558
  10:        0x105ae02d7 - core::panicking::panic::hb32bb819354fe8d9
  11:        0x104ba255b - hierarchical_pathfinding::path::generic_path::Path<P>::reversed::h585b2e252b320811
                               at /Users/patrick/.cargo/git/checkouts/hierarchical_pathfinding-4539d4dc23d64d2e/ae726be/src/path/generic_path.rs:48:19
  12:        0x104b986ee - hierarchical_pathfinding::path::path_segment::PathSegment::reversed::h4b3d583cf11002ec
                               at /Users/patrick/.cargo/git/checkouts/hierarchical_pathfinding-4539d4dc23d64d2e/ae726be/src/path/path_segment.rs:61:38
  13:        0x104ba5467 - hierarchical_pathfinding::graph::node_list::NodeList::add_edge::hf080584cc60fcd22
                               at /Users/patrick/.cargo/git/checkouts/hierarchical_pathfinding-4539d4dc23d64d2e/ae726be/src/graph/node_list.rs:50:26
  14:        0x1048dde9f - hierarchical_pathfinding::path_cache::PathCache<N>::tiles_changed_internal::h990089f5ffb31976
                               at /Users/patrick/.cargo/git/checkouts/hierarchical_pathfinding-4539d4dc23d64d2e/ae726be/src/path_cache.rs:1149:21
  15:        0x1048dbc44 - hierarchical_pathfinding::path_cache::PathCache<N>::tiles_changed::h8d9275fd8a51a1d5
                               at /Users/patrick/.cargo/git/checkouts/hierarchical_pathfinding-4539d4dc23d64d2e/ae726be/src/path_cache.rs:893:13
  16:        0x10493e98a - oligarchy::game::car::calculate_destination::calculate_destination::hed8733386fb0da34
                               at /Users/patrick/Projects/oligarchy/src/game/car/calculate_destination.rs:78:13
  17:        0x1049579e6 - core::ops::function::Fn::call::ha339ff97ae60f7a1
                               at /Users/patrick/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:70:5
```

**Comment by PatrickLerner** [comment]
Even more minimal, crashes every time:

```rust
use hierarchical_pathfinding::prelude::*;

const COST_MAP: [isize; 3] = [1, 10, -1];
type Grid = [[i32; 5]; 5];

// only borrows the Grid when called
fn cost_fn(grid: &Grid) -> impl '_ + FnMut((usize, usize)) -> isize {
    move |(x, y)| COST_MAP[grid[y as usize][x as usize] as usize]
}

fn main() {
    let mut grid = [
        [1, 1, 0, 0, 1],
        [0, 2, 1, 0, 1],
        [0, 0, 0, 0, 1],
        [1, 0, 0, 2, 0],
        [1, 1, 0, 2, 1],
    ];
    let (width, height) = (grid.len(), grid[0].len());

    let mut pathfinding = PathCache::new(
        (width, height),
        cost_fn(&grid),
        ManhattanNeighborhood::new(width, height),
        PathCacheConfig {
            chunk_size: 4,
            ..Default::default()
        },
    );

    grid[3][0] = 0;
    pathfinding.tiles_changed(&[(3, 0)], cost_fn(&grid));
}
```

**Comment by mich101mich** [comment]
In the example above there is an error where `cost_fn` maps `(x, y)` to `grid[y][x]`, but the change sets `grid[x][y]`. Those two need to be identical.

**Comment by PatrickLerner** [comment]
Thank you for the response :) That was not the problem actually in my project, but I actually figured it out:

My `cost_fn` is not really a calculated grid, but queries some state for objects that have a presence in a specific location. The problem was that when an object was despawned (which would have changed the `cost_fn` value for the position it used to be in), it did not emit the update to the pathfinding. When then some neighboring grid did update afterwards, it crashed.

So basically there were some updates not given to the pathfinding. When all updates are given to it, it works.
